### DISCUSSION TITLE
Problems with 0-1 decision variables in matlab gurobi?

### DISCUSSION THREAD

**Discussion by Electricman23** [yalmip/YALMIP#1175] [Optimization modelling]
Thanks for your help half a month ago, my program has improved a lot. 
At present, my understanding of 0-1 decision variables is very problematic.
First of all, I want to know whether in programming, I can define and assign values to 0-1 decision variables; 
Secondly, for example, in my program there is a 0-1 decision variable v(i,m), if at m=1, v(i,1)=1, then the subsequent time steps of variable v are 1 Can it be implemented according to the statement in my program;
```
 for m=2:NT
      for i=1:NC
         st=st+[implies(v(i,m-1),v(i,m)==1)];
     end
 end

for m=2:NT
     for i=1:NC
        st=st+[v(i,m-1)<=v(i,m)];
     end
end

st=st+[v(:,1)==v(:,2);
       v(:,3)==v(:,4);
       v(:,5)==v(:,6);
       v(:,7)==v(:,8);
       v(:,9)==v(:,10);
       ];
```
Finally, I would like to ask how you can look at results other than optimal.
My complete program looks like this:
```
warning('off');
clear
clc
yalmip;

%新能源场站内节点上的辅机及新能源机组信息
%第1列：节点号  第2列：辅机功率  第3列：额定容量  第4列：功频静特性系数（机组）  第5列：功频静特性系数（辅机）  第6列：惯性时间常数
Ebus=[
1,18000,2000000,25,1,10;
2,18000,2000000,25,1,10;
3,18000,2000000,25,1,10;
4,18000,2000000,25,1,10; 
5,18000,2000000,25,1,10;
6,18000,2000000,25,1,10;
7,18000,2000000,25,1,10;
8,18000,2000000,25,1,10;
9,18000,2000000,25,1,10;
10,18000,2000000,25,1,10;
11,18000,2000000,25,1,10;
12,18000,2000000,25,1,10;
13,18000,2000000,25,1,10;
14,18000,2000000,25,1,10;
15,18000,2000000,25,1,10;
16,18000,2000000,25,1,10;
17,18000,2000000,25,1,10;
18,18000,2000000,25,1,10;
19,18000,2000000,25,1,10;
20,18000,2000000,25,1,10;
21,18000,2000000,25,1,10;
22,18000,2000000,25,1,10;
23,18000,2000000,25,1,10;
24,18000,2000000,25,1,10;
25,18000,2000000,25,1,10;
26,18000,2000000,25,1,10;
27,18000,2000000,25,1,10;
];
ND=length(Ebus);%有27个节点

%新能源场站内线路信息
%第1列：线路号  第2列：首节点  第3列：末节点  第4列：rij  第5列：xij
Ebranch=[1,0,1;
2,1,2;
3,2,3;
4,3,4;
5,4,5;
6,0,6;
7,6,7;
8,7,8;
9,8,9;
10,9,10;
11,0,11;
12,11,12;
13,12,13;
14,13,14;
15,14,15;
16,0,16;
17,16,17;
18,17,18;
19,18,19;
20,19,20;
21,20,21;
22,0,22;
23,22,23;
24,23,24;
25,24,25;
26,25,26;
27,26,27;
];
NC=length(find(0==Ebranch(:,2)));%有5组
NE=[5;5;5;6;6];

%% 额定常数
NT=2*3;%NC;%最大总时步为2倍分组数
fmax=0.2/50;%最大稳态频率偏差；0.2/50P.U.
ft=0.5;%最大暂态频率偏差；
R=0.5/50;%频率变化量最大值  0.5HZ/s
SG=2500000;%外部带电域或储能的额定容量  2.5MW
KG=50;%外部带电域或储能的功频静特性系数  50p.u.
HG=5;%外部带电域或储能的惯性时间常数  5s
dfG=2;
PL=18000;
SW=2000000;
KW=25;
HW=10;
dfW=4;

%% 决策变量
K=sdpvar(1,NT,'full');%每一个时步的功频静特性系数
H=sdpvar(1,NT,'full');%每一个时步的惯性时间常数
P1=sdpvar(1,NT,'full');%每一个时步辅机所需要的功率
P2=sdpvar(1,NT,'full');%每一个时步新能源场站的可用容量
P3=sdpvar(1,NT,'full');%每一个时步新能源场站的出力
T=sdpvar(1,NT,'full');%每一时步的时间
t=sdpvar(NC,NT,'full');%每一组辅机或者风机恢复需要的时间
n=sdpvar(NC,NT,'full');%每一组已恢复的辅机数
k=sdpvar(NC,NT,'full');%每一组已恢复的风机数
Nn=sdpvar(1,NT,'full');
Nk=sdpvar(1,NT,'full');
v=binvar(NC,NT,'full');%每一个时步每一组新能源机组辅机的运行状态
u=binvar(NC,NT,'full');%每一个时步每一组新能源机组的运行状态   0表示未稳定出力  1表示稳定出力
a=binvar(11,NT,'full');


%% 目标函数
st=[];
% f=[];
for m=1:NT
                f=-sum(P3(1,m)*T(1,m));
end

%为解决目标无界性，对所有决策变量进行约束
% for m=1:NT
%     for i=1:NC
        st=st+[0<=t<=100;
               0<=n<=10;
               0<=k<=10];
%     end
% end

% for m=1:NT
        st=st+[0<=K<=100;
               0<=H<=100;
               0<=P1<=100*PL;
               0<=P2<=100*SW;
               40<=T<=100;
               0<=Nn<=100;
               0<=Nk<=100;
               ];
% end

%% 约束条件
%根据约束对第一时步相关参数进行预设
for m=1
        st=st+[
%                u(1:5,m)==0;
               K(1,m)==KG;
               H(1,m)==HG;
               P1(1,m)==90000;
               P2(1,m)==0;
               ];
end

%0/1决策变量约束，辅机的状态量与新能源机组的状态量间的约束
%偶数时步决策变量v不会变化，奇数时步会选择v是否为0/1
% for m=2:NT
%      for i=1:NC
%         st=st+[implies(v(i,m-1),v(i,m)==1)];
%      end
% end

for m=2:NT
     for i=1:NC
        st=st+[v(i,m-1)<=v(i,m)];
     end
end

st=st+[v(:,1)==v(:,2);
       v(:,3)==v(:,4);
       v(:,5)==v(:,6);
%        v(:,7)==v(:,8);
%        v(:,9)==v(:,10);
       ];

%u与v的关系
for m=2:NT
     for i=1:NC
        st=st+[u(i,m)==v(i,m-1)];%辅机启动后下一时步新能源机组启动
     end
end

%对每一时步的01决策变量进行计数
for m=1:NT
      for i=1:NC
            st=st+[n(i,m)==v(i,m)*NE(i,1);
                   k(i,m)==u(i,m)*NE(i,1)];
      end
end
%当前已恢复的辅机和风电机组总数
st=st+[Nn==sum(n,1);
       Nk==sum(k,1)];

% 时间约束    新能源辅机恢复时，新能源所在线路及变压器恢复考虑在这一时步（40+20+5*n）新能源机组恢复时（40+10）
for m=1
    for i=1:NC
        t(i,m)=60*v(i,m)+5*n(i,m)+50*u(i,m);
    end
end

for m=2:NT
    for i=1:NC
        t(i,m)=60*(v(i,m)-v(i,m-1))+5*(n(i,m)-n(i,m-1))+50*(u(i,m)-u(i,m-1));
    end
end

Z=40*ones(1,NT);
t1=[t;Z];
st=st+[T==max(t1,[],1)];%每一个时步的时间按照每组恢复的最大时间计算

%工频静特性系数以及惯性时间常数
for m=2:NT
        st=st+[sum(a,1)==1;
               implies(a(1,m),[Nk(1,m)==5,K(1,m)==(5*SW*KW+SG*KG)/(5*SW+SG),H(1,m)==(5*SW*HW+SG*HG)/(5*SW+SG)]);
               implies(a(2,m),[Nk(1,m)==6,K(1,m)==(6*SW*KW+SG*KG)/(6*SW+SG),H(1,m)==(6*SW*HW+SG*HG)/(6*SW+SG)]);
               implies(a(3,m),[Nk(1,m)==10,K(1,m)==(10*SW*KW+SG*KG)/(10*SW+SG),H(1,m)==(10*SW*HW+SG*HG)/(10*SW+SG)]);
               implies(a(4,m),[Nk(1,m)==11,K(1,m)==(11*SW*KW+SG*KG)/(11*SW+SG),H(1,m)==(11*SW*HW+SG*HG)/(11*SW+SG)]);
               implies(a(5,m),[Nk(1,m)==12,K(1,m)==(12*SW*KW+SG*KG)/(12*SW+SG),H(1,m)==(12*SW*HW+SG*HG)/(12*SW+SG)]);
               implies(a(6,m),[Nk(1,m)==15,K(1,m)==(15*SW*KW+SG*KG)/(15*SW+SG),H(1,m)==(15*SW*HW+SG*HG)/(15*SW+SG)]);
               implies(a(7,m),[Nk(1,m)==16,K(1,m)==(16*SW*KW+SG*KG)/(16*SW+SG),H(1,m)==(16*SW*HW+SG*HG)/(16*SW+SG)]);
               implies(a(8,m),[Nk(1,m)==17,K(1,m)==(17*SW*KW+SG*KG)/(17*SW+SG),H(1,m)==(17*SW*HW+SG*HG)/(17*SW+SG)]);
               implies(a(9,m),[Nk(1,m)==21,K(1,m)==(21*SW*KW+SG*KG)/(21*SW+SG),H(1,m)==(21*SW*HW+SG*HG)/(21*SW+SG)]);
               implies(a(10,m),[Nk(1,m)==22,K(1,m)==(22*SW*KW+SG*KG)/(22*SW+SG),H(1,m)==(22*SW*HW+SG*HG)/(22*SW+SG)]);
               implies(a(11,m),[Nk(1,m)==27,K(1,m)==(27*SW*KW+SG*KG)/(27*SW+SG),H(1,m)==(27*SW*HW+SG*HG)/(27*SW+SG)]);
               ]; 
end

% 稳态频率约束
for m=2:NT
%     for i=1:NC
        st=st+[
%                S1(1,m)==SW*Nk(1,m)+SG;
%                S2(1,m)==KW*SW*sum(k(i,m))+KG*SG;
%                K(1,m)==S2(1,m)/S1(1,m);%H=A/B→H*B=A无法解决，采用implies全部列举出来
               P1(1,m)<=fmax*K(1,m)*(SW*Nk(1,m)+SG);
               P2(1,m)<=fmax*K(1,m)*(SW*Nk(1,m)+SG)
               ];
%     end
end

% 惯量响应约束
for m=2:NT
%     for i=1:NC
        st=st+[
%                S1(1,m)*H(1,m)==HW*SW*sum(k(i,m))+HG*SG;%H=A/B→H*B=A
               P1(1,m)<=2*R*H(1,m)*(SW*Nk(1,m)+SG);
               P2(1,m)<=2*R*H(1,m)*(SW*Nk(1,m)+SG)
               ];
%     end
end

%暂态频率约束
for m=2:NT
%     for i=1:NC
        st=st+[
%                S1(1,m)*H(1,m)==HW*SW*sum(k(i,m))+HG*SG;%H=A/B→H*B=A
               P1(1,m)<=ft*(Nk(1,m)*SW/dfW+SG/dfG);
               P2(1,m)<=ft*(Nk(1,m)*SW/dfW+SG/dfG)
               ];
%     end
end

% 启动功率以及新能源机组出力等式关系
for m=1:NT
%     for i=1:NC
        st=st+[P1(1,m)==PL*Nn(1,m);
%              P2(1,m)==0.1*SW*sum(k(i,m))
               0.01*SW*Nk(1,m)<=P2(1,m)<=SW*Nk(1,m);
               P3(1,m)==SW*Nk(1,m)];
%     end
end

%% 求解
gurobi_stop_gap=10^(-4); % 设置求解混合整数规划的 Gap 为10^(-4)
longest_solve_time=300; % 设置最长求解时间为 300 秒
NumericFocus=2;%
options=sdpsettings('verbose',2,'solver','gurobi+', 'debug', 1, 'gurobi.NonConvex', 2,'gurobi.timelimit',longest_solve_time,...
    'gurobi.MIPGap', gurobi_stop_gap);%cplex%,'gurobi.NumericFocus',NumericFocus
result=optimize(st,f,options)
value(f);

%% 结论
AP1=value(P1);
AP2=value(P2);
Bv=value(v);
Bu=value(u);
BK=value(K);
BH=value(H);
Cn=value(n);
Ck=value(k);
C1=value(Nn);
C2=value(Nk);
CT=value(T);
D=value(t1);
D1=value(a);
```

**Comment by johanlofberg** [comment]
It is not clear what you are asking with define and assign. If you want some variables to have a value, simply add the constraint x(i)==value (or just replace the variable with a value, x(i)=value) Using constraints typically leads to more clear code and less oppurtunities for bugs/strange phenomena when you are mixing decision variables and constants in your vector variables

If you want x to be true when y is true, you do implies(x,y) which simply ends up as y>=x for binary variables

**Reply by Electricman23** [reply]
First of all, the first question: what I want to ask is whether the 0-1 variable can be given a numeric value to v. Or as I showed, in the implies statement, after the v of the current moment is 1, the v of all subsequent moments is 1. 
```
for m=2:NT
     for i=1:NC
        st=st+[implies(v(i,m-1),v(i,m)==1)];
     end
end
```
The second problem: how to make my 0-1 decision variable v consistent with the first moment at moment 2, that is, the v at moment 2 does not act, whether the program I am showing is reasonable.
```
st=st+[v(:,1)==v(:,2);
       v(:,3)==v(:,4);
       v(:,5)==v(:,6);
       v(:,7)==v(:,8);
       v(:,9)==v(:,10);
       ];
```

**Reply by Electricman23** [reply]
Another problem is that after I program according to the current logic, the results of the programming are obviously unreasonable. Suppose I have a total of 10 operation time steps, according to common sense, the fifth time step can make all 0-1 decision variables v become 1, but the program always shows that the last time step v all become 1.
I would like to ask if this happens because of an error in my logic or if there is a problem with my objective function.
Finally, how should I look at other results than optimal results.

**Reply by johanlofberg** [reply]
As I said, if you want y, i.e v(i,m), to be true when x, i.e v(i,m-1), is true, you do implies(x,y) which is equivalent to y>=x

The second question I don't know what you are asking. You are currently simply saying that v(1) and v(2) are always are the same etc (meaning implies(v(m-1),v(m)) is redundant for even m, although it doesn't affect anything to keep those for simplicity)

**Reply by johanlofberg** [reply]
Add a constraint where you force the solution to have some structure which you claim is more reasonable. If it is infeasible, well you've got an issue in the model or in your thinking and must debug what causes that infeasibility. If it is feasible but with higher objective, well same thing.

**Reply by Electricman23** [reply]
For the second question, I will explain it as follows: Suppose there are n groups of motors that need to restore power supply, and each group has a state decision variable v of the auxiliary machine and a state decision variable u of the output. Only when the decision variable v of the auxiliary machine becomes 1, the motor will contribute, and the decision variable u of the motor will become 1. Suppose there are a total of 10 time steps, the decision variable v of the odd time step auxiliary machine changes, and the even time step motor output decision variable u changes. For example, in step 1, the auxiliary decision variable v changes, but the v of step 2 should be consistent with the v of step 1, which is why the state quantities of 1 and 2 hour steps in my program are consistent.

**Reply by johanlofberg** [reply]
All your code is pairing variables, and then saying that a pair which is true forces all following pairs to be true, i.e.the only feasible solutions are

0 0 0 0 0 0 0 0 0 ...
0 0 1 1 1 1 1 1 1 ...
0 0 0 0 1 1 1 1 1...

**Reply by Electricman23** [reply]
Sorry, I didn't understand what you meant. Can I understand that if you write a program according to my logic, you will have the current erroneous result.

**Reply by johanlofberg** [reply]
All am am saying is that the implies and equalities on the v(m) variables that you posted in the first post simply imply a pairing and a "no-turn-off" where the only feasible solutions are those I listed.

**Reply by Electricman23** [reply]
Do you mean the solution that can be used to deal with the problem, or does it refer to the result of the program running?

**Reply by johanlofberg** [reply]
Am am just saying what kind of solutions you ever can get when you have those constraints, i.e. what you are encoding in case you think you are encoding something else.

**Reply by Electricman23** [reply]
Oh, okay. I understand what you mean. As you said, my viable solution does.

**Reply by Electricman23** [reply]
The results are as follows: (5 groups in total)
0 0 0 0 0 0 0 0 1 1
0 0 0 0 0 0 1 1 1 1
1 1 1 1 1 1 1 1 1 1
0 0 0 0 1 1 1 1 1 1
0 0 1 1 1 1 1 1 1 1
However, according to my constraints, when I am in the fourth step, I can revert the v of groups 4 and 5 to 1, but the current program represents a set that seems to be restored at a time.

**Reply by Electricman23** [reply]
Can I understand that if I don't add a constraint to force him to get a reasonable scheme, then he defaults to the last time-step decision variable v all becomes 1.

**Reply by johanlofberg** [reply]
Don't understand what you mean with revert groups of 4 and 5 to 1

If you don't add a constraint, there is no default value. The solver picks what ever is best from the objective, and if it does not influence the objective, it can be anything

**Reply by Electricman23** [reply]
The explanation is as follows:

The result of the program running：
0 0 0 0 0 0 0 0 1 1
0 0 0 0 0 0 1 1 1 1
1 1 1 1 1 1 1 1 1 1
0 0 0 0 1 1 1 1 1 1
0 0 1 1 1 1 1 1 1 1 

Desirable results：
0 0 0 0 1 1 1 1 1 1
0 0 0 0 1 1 1 1 1 1
1 1 1 1 1 1 1 1 1 1
0 0 1 1 1 1 1 1 1 1
0 0 1 1 1 1 1 1 1 1 

The program will not be all 1s until step 9, and the ideal result will be all 1s in the middle of the time step.

**Reply by johanlofberg** [reply]
As I said, add the constraint that you want this solution. If it is infeasible, you've got some serious issue in your model or idea of how the system works. If it is feasible but with an higher objective, you misunderstood something that causes this to be a worse solution. If you get same objective, it means it does not make a difference in terms of the objective, and if you want the solver to pick this solution, you must add something to the model to tell the solver that this is a better solution

**Reply by Electricman23** [reply]
Okay, I'll make adjustments as you suggest. But I also want to ask you a question, can I see all the solutions in the solver, including those that are not optimal, not just the optimal solution. Because, when I look at the results, I find that the solver represents very few results.
```
Explored 0 nodes (0 simplex iterations) in 0.13 seconds (0.01 work units)
Thread count was 4 (of 4 available processors)

Solution count 3: -2.7e+09 -1.6e+09 -1.5e+09 
No other solutions better than -2.7e+09

Optimal solution found (tolerance 1.00e-04)
Best objective -2.700000000000e+09, best bound -2.700000000000e+09, gap 0.0000%
```

**Reply by johanlofberg** [reply]
All that log is saying is that the problem was utterly trivial to solve. Returning all solutions is typically not possible, what you can do is to resolve with a constraint that the previous solution is not allowed (which can be done with the operator exclude) and keep running until infeasible

**Reply by Electricman23** [reply]
Okay, I get it. I also wanted to know if all decision variables should be written into the constraint. For example, if I sum a continuous decision variable in which a 0-1 decision variable is multiplied by a constant term, then whether the sum term is written in the form of a constraint.
```
for m=1:NT
      for i=1:NC
            st=st+[n(i,m)==v(i,m)*NE(i,1);
                   k(i,m)==u(i,m)*NE(i,1)];
      end
end

st=st+[Nn==sum(n,1);
       Nk==sum(k,1)];
```
That is, the Nk and whether Nn is to be written as a constraint.

**Reply by johanlofberg** [reply]
Looks like you are doing some messy thing to simply assign Nn = sum(v.*repmat(Ne,1,NC))

**Reply by Electricman23** [reply]
If I modify it as you write it, should the equation be a constraint?

**Reply by johanlofberg** [reply]
No, that is simply an assignment (=, not ==). No reason to define decision variables and define them via constraints.

**Reply by Electricman23** [reply]
But Nk is the value associated with the 0-1 decision variable, is it variable, shouldn't it be defined as a decision variable, or fill in the resulting value into a matrix?

**Reply by johanlofberg** [reply]
This model has one decision variable, x

```
sdpvar x
y = 1+x;
optimize([],x^2+y^2)
```

This model has two decision variables, x and y

```
sdpvar x y
optimize([y==1+x],x^2+y^2)
```

**Reply by Electricman23** [reply]
I think I should have understood what you meant. What you're trying to say is that if I define y as a decision variable with sdpvar, then y should be in the constraint?

**Reply by Electricman23** [reply]
For example
```
binvar v
sdpvar Nk
objective=[]
constraint=[Nk==sum(v,1)]
optimize(constraint,objective)
```
I don't know if you can understand the program I described above, and if I write it this way, is it not a mistake?

**Reply by johanlofberg** [reply]
If you first define y as an sdpvar and then assign it an expression, the initial definition of y as an sdpvar would have been redundant code. If you intend to use y in a constraint it must first have been defined as an sdpvar so it exist (or create through an assignment, but then you are implicitly adding a constraint on some other variable in that constraint).

**Reply by Electricman23** [reply]
Can I understand that I can define and assign values in this way, and that y in the constraint requires a sdpvar definition.

**Reply by johanlofberg** [reply]
```
sdpvar x
y = x+1
Model = [y == 5]
optimize(Model, x^2+y^2)
```

is exactly the same as (solver will see absolutely the same thing)

```
sdpvar x
Model = [x+1 == 5]
optimize(Model, x^2+(x+1)^2)
```

and is from a modelling perspective same thing as (but now with a different model in two variables so solver solves another problem)

```
sdpvar x y
Model = [y == x+1, y == 5]
optimize(Model, x^2+y^2)
```
The following code does not run, as y is used without being declared or defined

```
sdpvar x
Model = [y == x+1, y == 5]
optimize(Model, x^2+y^2)

```

**Comment by Electricman23** [comment]
Hello johanlofberg. Sorry to bother you. I still have two questions to ask.
The first is whether it is possible to multiply two consecutive decision variables in the objective function. 
The following procedure describes the equation relationships for P3 and T in the objective function, respectively.
```
P3=sdpvar(1,NT,'full');
T=sdpvar(1,NT,'full');

for m=1:NT
                f=-sum(P3(1,m)*T(1,m));
end

for m=1:NT
               P3(1,m)==0.5*SW*Nk(1,m)];
end

for m=1
    for i=1:NC
        t(i,m)=60*v(i,m)+5*n(i,m)+50*u(i,m);
    end
end

for m=2:NT
    for i=1:NC
        t(i,m)=60*(v(i,m)-v(i,m-1))+5*(n(i,m)-n(i,m-1))+50*(u(i,m)-u(i,m-1));
    end
end

Z=40*ones(1,NT);
t1=[t;Z];
st=st+[T==max(t1,[],1)];
```
The second is the Nk formula problem that we talked about yesterday, what are the consequences of taking the complex method I wrote, and whether it will affect the results of my programming.
```
n=sdpvar(NC,NT,'full');
k=sdpvar(NC,NT,'full');
Nn=sdpvar(1,NT,'full');
Nk=sdpvar(1,NT,'full');
v=binvar(NC,NT,'full');
u=binvar(NC,NT,'full');

for m=1:NT
      for i=1:NC
          n(i,m)=v(i,m)*NE(i,1);
          k(i,m)=u(i,m)*NE(i,1);
      end
end

st=st+[Nn==sum(n,1);
       Nk==sum(k,1)];

NE=[5;5;5;6;6;6];
```

**Reply by Electricman23** [reply]
There is also the result you mentioned yesterday about the result that I mentioned after setting the relevant constraint (u v).

0 0 0 0 0 0 0 0 0 ...
0 0 1 1 1 1 1 1 1 1 ...
0 0 0 0 1 1 1 1 1 1...

I would like to ask if I follow the program I have written, and the result is that the odd-numbered time step restores only one line, not 2-3 lines as I thought

**Reply by Electricman23** [reply]
I made changes to whether some equations were added to the constraints while debugging the program and wondered about it. 
For example, time is due to the 0-1 variable, and when I take it as a constraint, I get the following result. 
1.Does not include a program segment and does not add constraints
1	1	1	1	0	0	1	1	1	1 ...
0	0	1	1	0	0	1	1	1	1 ...
0	0	1	1	0	0	1	1	1	1 ...
0	0	0	0	1	1	0	0	1	1 ...
0	0	1	1	0	0	1	1	1	1 ...

Without including constraints, another result is obtained.
2.Does not include a program segment but adds a constraint
0	0	1	1	1	1	1	1	1	1
0	0	1	1	1	1	1	1	1	1
1	1	1	1	1	1	1	1	1	1
0	0	0	0	0	0	0	0	1	1
0	0	0	0	0	0	0	0	1	1

And when I add the following block, the result becomes the following result.
```
for m=2:NT
     for i=1:NC
        st=st+[implies(v(i,m-1),v(i,m)==1)];
     end
end
```
3.Includes a program segment and adds a constraint
0	0	0	0	0	0	0	0	1	1
0	0	0	0	0	0	1	1	1	1
1	1	1	1	1	1	1	1	1	1
0	0	0	0	1	1	1	1	1	1
0	0	1	1	1	1	1	1	1	1

So I want to know what the equation in the constraint should look like?

**Reply by johanlofberg** [reply]
This

```
for m=1:NT
                f=-sum(P3(1,m)*T(1,m));
end
```

makes no sense as the loop doesn't keep any intermediate results, and the sum is redundant, as you simply can replace everything with

`f=-P3(1,NT)*T(1,NT);`

Products of decision variables leads to bilinear expressions, which leads to indefinite quadratic programming which is orders of magnitudes harder, although gurobi supports it.

I have no idea what you are talking about when you talk about having added constraints or not, and then showing solutions or desired solutions or whatever it is

**Reply by Electricman23** [reply]
Let me ask if my objective function can only be expressed in its current form, is there a way to change its expression?
As for the question I asked above about whether the related equation should be added to the constraint, I want to know what kind of situations need to be added to the constraint, and whether all equations and inequalities need to be added to the constraint. Because during the process of program modification, I found that when some equations are not added to the constraint, the solution will have different results.

**Reply by johanlofberg** [reply]
I have no idea what you mean with change it or expressed in current form. At the moment your objective is a single bilinear term (if it is f you are talking about). Most likely you forgotten to accumulate the parts when running the loop and intend to have a sum of bilinear terms. Bilinear objective is haaard, but there is no way around it if that is what you want.

Still have no idea what you are talking about concerning the constraints. If you want the solution to have a particular structure, you have to encode that in the constraints. If you want x to be true when y is true, then you must include x>=y (or implies(x,y) which you use). There is nothing more going on in your code.

**Reply by Electricman23** [reply]
You mean, my objective function can't be bypassed or avoided because it's necessary for me, but it's possible to solve it with the gurobi solver, right?
As for constraints, I'll illustrate them with examples from my program.
Such as the following procedure:
```
for m=1
    for i=1:NC
        t(i,m)=60*v(i,m)+5*n(i,m)+50*u(i,m);
    end
end

for m=2:NT
    for i=1:NC
        t(i,m)=60*(v(i,m)-v(i,m-1))+5*(n(i,m)-n(i,m-1))+50*(u(i,m)-u(i,m-1));
    end
end

Z=40*ones(1,NT);
t1=[t;Z];
st=st+[T==max(t1,[],1)];
```
The above procedure is about solving for time T.where t is the time calculation for each row, and T is the maximum value filtered out. So whether t and T should I add it to the constraint in the form st=st+[].

**Reply by johanlofberg** [reply]
I haven't said you cannot use bilinear objective, I only said it leads to a much harder problems. To make things more confusing, you haven't even explicitly told us what term is your objective.

Your code is unnecessarily complicated also. It looks to me like like you are simply defining

```
t = 60*v+5*n+50*u
t = [t(:,1) diff(t,1)]

```
or something like that

**Reply by Electricman23** [reply]
Yes, I checked it out and found that what I was describing was the form you were describing, but I was describing it more complexly.
 For the description of the objective function, I will make the following explanation, I hope you can understand.
Power generation f = sum (output power P3(1,m) × per period of time T(1,m))

**Reply by Electricman23** [reply]
Therefore, what I also want to ask is whether I should add the equation description of time T to the constraint.

**Reply by johanlofberg** [reply]
I have no idea if you should add a constraint, as it is you who know what must hold or not

**Comment by Electricman23** [comment]
Hello, I recently extended my objective function. Because according to the model I built, I mean that there are three decision variables in my objective function, namely
```
sdpvar x y z
f=-x*y/z
```
Is it possible to multiply and divide the three decision variables like the one represented above?

**Reply by johanlofberg** [reply]
Well, it works as you obviously did create the objective. The problem is that you have to realize that you probably will have to use a general nonlinear solver for the problem (unless you are lucky and it turns out to be a geometric program. the previous model you showed has no chance of being a GP, and as you had binary varables in it, it appears completely intractable )

**Reply by Electricman23** [reply]
So, can the gurobi solver I use accomplish this goal or the constraints of that type?

**Reply by johanlofberg** [reply]
no, since gurobi only deals with mixed integer quadratic stuff. I would say there is no solver capable of solving this beyond trivially small examples

**Reply by Electricman23** [reply]
Well, if my constraints don't have to be met at every point in time, then do I need to separate the constraints for each time step separately?

**Reply by johanlofberg** [reply]
Of course

**Reply by Electricman23** [reply]
Well, I would also like to ask about the question of when to use invar and the difference between invar and sdpvar.

**Reply by johanlofberg** [reply]
intvar is just an sdpvar with the added constraint that it is integer

**Comment by Electricman23** [comment]
Hello, I would like to consult you on a very basic question, I hope you do not give up. That is, the question of decision variables, that is, when to use sdpvar. Whether all the variables I should make him a decision variable. For example, if x is a 0-1 decision variable, and y = 5 * x, then whether y should be a decision variable.

**Reply by johanlofberg** [reply]
All you are saying is, hey lets simplify my notation here and give the term 5x the name y. They are both built from the same basic variable

```
>> yalmip('clear')
>> x = binvar(1);
>> z = binvar(1);
>> y = 5*x;
>> see(x)
 
Constant matrix
 
     0

Base matrices
 
     1

 
Used variables
 
     1

>> see(y)
 
Constant matrix
 
     0

Base matrices
 
     5

 
Used variables
 
     1

>> see(z)
 
Constant matrix
 
     0

Base matrices
 
     1

 
Used variables
 
     2

>> see([x y;3 z])
 
Constant matrix
 
     0     0
     3     0

Base matrices
 
     1     5
     0     0

 
     0     0
     0     1

 
Used variables
 
     1     2


```

**Reply by Electricman23** [reply]
But when I put y into a constraint or into a target function, if I don't define sdpvar y, the program represents it as undefined y.

**Reply by johanlofberg** [reply]
That simply means you haven't made any assignment such as y=5 * x before referencing y in your code

Soundes like you've done something like this, which makes no sense

```
sdpvar x
Model = [y == x]
```
Since y neither has been created as a variable, nor has it been defned as some function of the available variables (i.e. x), it just doesn't exist when it is referenced

**Reply by Electricman23** [reply]
That is, if y is used in my goal or constraint, and my y is constructed from the decision variable x, i.e. y = 5* x, sdpvar x. then I still need to define sdpvar y. Is that what you mean?

**Reply by johanlofberg** [reply]
Once again, you either create variables as new decision variables and then use them, or you temporarily create expressions from other decision variables and expressions and give them some name

If you want to minimize x + (x+1) + (x+1)^2 and you want to be lazy in typing, you either introduce a temporary placeholder for the expression x+1 and solve a problem in dimension 1

```
sdpvar x
y = x+1
objective = x+y+y^2;
model = [];
optimize(model,objective)

```
or you introduce a new variable and solve a problem in dimension 2 with a constraint which implicitly means it is solved in a 1-dimensional subspace

```
sdpvar x y
Model = [y == x+1];
objective = x+y+y^2;
optimize(model,objective)
```

This won't work since there is nothing called y when it is referenced

```
sdpvar x
Model = [y == x+1];
objective = x+y+y^2;
optimize(model,objective)
```

Neither will this

```
sdpvar x
objective = x+y+y^2;
y = x+1
optimize(model,objective)
```

And this is just redundant since the initial definition of y as an independent variable is lost when it is reassigned as an expression built from x

```
sdpvar x y
y = x+1
objective = x+y+y^2;
optimize(model,objective)
```
